### PR TITLE
Fix chat bug and decluttering tools in forms

### PR DESCRIPTION
**Pull request by timbl**
- Add ability to suppress form fields when the data is read-only. 
 - For security, add the fourth component to kb.each() etc .
 - Fix the bug that finite chat goes crazy loading data and displays nothing if it runs before the div is in place.

### CODE REVIEW

File: src/style.js [hunk 1]
```diff
@@ -18,16 +18,26 @@ module.exports = {
...
+  personaBarStyle: 'width: 100%; height: 4em; background-color: #eee; vertical-align: middle;',
   // Login buttons
 
   signInButtonStyle: 'padding: 1em; border-radius:0.2em; margin: 2em; font-size: 100%;', // was 0.5em radius
   // Forms
 
+  heading1Style: 'font-size: 180%; font-weight: bold; color: #888888; padding: 0.5em;', // originall was brown now grey
+  heading2Style: 'font-size: 130%; font-weight: bold; color: #888888; padding: 0.4em;', // originall was brown now grey
+  heading3Style: 'font-size: 120%; font-weight: bold; color: #888888; padding: 0.3em;', // For eg in large forms or before a small form
+  heading4Style: 'font-size: 110%; font-weight: bold; color: #888888; padding: 0.2em;', // Lowest level used by default in smalll things
```
Comment (TallTed) [inline src/style.js:31]:
```suggestion
  heading1Style: 'font-size: 180%; font-weight: bold; color: #888888; padding: 0.5em;', // originally was brown; now grey
  heading2Style: 'font-size: 130%; font-weight: bold; color: #888888; padding: 0.4em;', // originally was brown; now grey
  heading3Style: 'font-size: 120%; font-weight: bold; color: #888888; padding: 0.3em;', // For example, in large forms or before a small form
  heading4Style: 'font-size: 110%; font-weight: bold; color: #888888; padding: 0.2em;', // Lowest level used by default in small things
```

File: src/style.js [hunk 2]
```diff
@@ -18,16 +18,26 @@ module.exports = {
...
+
   formBorderColor: '#888888', // originall was brown now grey
   formHeadingColor: '#888888', // originall was brown now grey
   formHeadingStyle: 'font-size: 110%; font-weight: bold; color: #888888; padding: 0.2em;', // originall was brown now grey
```
Comment (TallTed) [inline src/style.js:35]:
```suggestion
  formBorderColor: '#888888', // originally was brown; now grey
  formHeadingColor: '#888888', // originally was brown; now grey
  formHeadingStyle: 'font-size: 110%; font-weight: bold; color: #888888; padding: 0.2em;', // originally was brown; now grey
```

File: src/widgets/buttons.ts [hunk 3]
```diff
@@ -405,7 +405,7 @@ export function setImage (element: HTMLElement, thing: NamedNode) { // 20191230a
 
 // If a web page then a favicon with a fallback to
 // See eg http://stackoverflow.com/questions/980855/inputting-a-default-image
```
Comment (TallTed) [inline src/widgets/buttons.ts:407]:
```suggestion
// If a web page, then a favicon, with a fallback to ???
// See, e.g., http://stackoverflow.com/questions/980855/inputting-a-default-image
```

File: src/widgets/forms/basic.ts [hunk 4]
```diff
@@ -101,6 +101,9 @@ export function basicField (
     )
     return box
   }
+  // It can be cleaner to just remove empty feilds if you can't edit them anyway
```
Comment (TallTed) [inline src/widgets/forms/basic.ts:104]:
```suggestion
  // It can be cleaner to just remove empty fields if you can't edit them anyway
```

**Review by TallTed — CHANGES_REQUESTED** [review]
human-facing text edits in code comments

File: src/chat/infinite.js [hunk 5]
```diff
@@ -787,6 +787,16 @@ export async function infiniteMessageArea (dom, kb, chatChannel, options) {
       !earliest.messageTable.initial &&
       earliest.messageTable.extendBackwards
     ) {
+      // If this has been called before the element is actually in the
+      // user's DOM tree, then this scrollTop check won't work -> loop forever
```
Comment (angelo-v) [inline src/chat/infinite.js:791]: **question:** I don't get it from the comment yet: Is the "loop forever" fixed by these lines, or is there still an issue left?

File: src/widgets/forms.js [hunk 6]
```diff
@@ -1183,7 +1197,7 @@ forms.makeDescription = function (
   }
   function saveChange (_e) {
     submit.disabled = true
-    submit.setAttribute('style', 'visibility: hidden; float: right;') // Keep UI clean
+    submit.setAttribute('style', ' : hidden; float: right;') // Keep UI clean
```
Comment (angelo-v) [inline src/widgets/forms.js:1200]: **question**: Is this intentional? I wonder why `' : hidden;` is left without a preceding css property

File: src/chat/infinite.js (hunk 5 above)
Comment (timbl) [reply src/chat/infinite.js:791]: Its fixed by spotting that scrolHeight is zero, and putting off the load of the chat till 2 secs later

File: src/widgets/forms.js (hunk 6 above)
Comment (timbl) [reply src/widgets/forms.js]: Fixed in latest commit

**Pull request merged**